lib/jxl/modular/encoding/enc_encoding.cc

```diff
@@ -479,24 +485,42 @@ Status ModularEncode(const Image &image, const ModularOptions &options,
   }
 
   size_t image_width = 0;
+  size_t total_tokens = 0;
   for (size_t i = 0; i < nb_channels; i++) {
-    if (!image.channel[i].w || !image.channel[i].h) {
-      continue;  // skip empty channels
-    }
     if (i >= image.nb_meta_channels &&
         (image.channel[i].w > options.max_chan_size ||
          image.channel[i].h > options.max_chan_size)) {
       break;
     }
     if (image.channel[i].w > image_width) image_width = image.channel[i].w;
-    if (options.zero_tokens) {
-      tokens->resize(tokens->size() + image.channel[i].w * image.channel[i].h,
-                     {0, 0});
-    } else {
+    total_tokens += image.channel[i].w * image.channel[i].h;
+  }
+  if (options.zero_tokens) {
+    tokens->resize(tokens->size() + total_tokens, {0, 0});
+  } else {
+    // Do one big allocation for all the tokens we'll need,
+    // to avoid reallocs that might require copying.
+    // Also make the allocation non-initializing.
+    size_t pos = tokens->size();
+    auto nia_tokens =
+        reinterpret_cast<std::vector<Token, no_init_alloc<Token>> *>(tokens);
```

Comment:
Uhhhh... no. This is *definitely* UB.

What's the speed difference wrt a good old reserve()?

Reply:
Well reserve() doesn't help because then afterwards you still need to do emplace_back() (capacity check for every token) or resize() (which will unnecessarily initialize the memory). Doing an initializing resize() is already better than having a capacity check per token, but it's still a few percent slower than avoiding that unnecessary initialization.

I suppose the cleaner way to do it is to define a type `std::vector<Token, no_init_alloc<Token>>` and use it consistently instead of a vector with the default allocator.

Reply:
Can you benchmark the various options (resize(), reserve(), this PR and current state)?

I'm not going to LGTM this change as it is. I would be OK with the unallocated vector if no reinterpret_cast is involved (and if the allocator doesn't introduce UB in interesting ways either)

Reply:
Yes, I agree that the reinterpret_cast is nasty. I think the best approach is to have our own class for Token vectors, which is basically just the same thing as std::vector<Token, no_init_allocator>. That'll take some significant code changes though (we have lots of stuff using std::vector<Token>) so maybe I'll change this PR to do an initializing resize and then later it will become a non-initialized resize.

Reply:
Updated it, the difference between this and skipping the initialization is small.

Reply:
Can you update the numbers above / add new ones?

Reply:
| effort | before | after (no init) | after (with init, current PR) |
|-------|--------|------|-------|
| 1 (lightning) | 25.91 MP/s | 36.57 MP/s | 36.79 MP/s |
| 2 (thunder) | 21.44 MP/s | 26.48 MP/s | 26.10 MP/s |
| 3 (falcon) | 14.55 MP/s | 17.59 MP/s | 17.47 MP/s |

The difference between no init and init is probably more noise than actual difference.

Reply:
Ok, so I propose we don't do anything further ;)